src/SQLite.jl

```diff
@@ -12,10 +12,44 @@ end
 include("consts.jl")
 include("api.jl")
 
+const DBHandle = Ptr{Cvoid}   # SQLite3 DB connection handle
+const StmtHandle = Ptr{Cvoid} # SQLite3 prepared statement handle
+
 # Normal constructor from filename
-sqliteopen(file, handle) = sqlite3_open(file, handle)
-sqliteerror(db) = throw(SQLiteException(unsafe_string(sqlite3_errmsg(db.handle))))
-sqliteexception(db) = SQLiteException(unsafe_string(sqlite3_errmsg(db.handle)))
+sqliteexception(handle::DBHandle) = SQLiteException(unsafe_string(sqlite3_errmsg(handle)))
+sqliteerror(handle::DBHandle) = throw(sqliteexception(handle))
+
+"""
+The wrapper that holds the handle to SQLite3 prepared statement.
+It is managed by [`SQLite.DB`](@ref) and referenced by "public" [`SQLite.Stmt`](@ref) objects.
+
+When no `SQLite.Stmt` instances reference the given `SQlite._Stmt` object
+(its `refcount` goes to 0), it is closed automatically.
+
+When `SQLite.DB` is closed or [`SQLite.finalize_statements!`](@ref) is called,
+all its `SQLite._Stmt` objects are closed.
+"""
+mutable struct _Stmt
+    handle::StmtHandle
+    params::Dict{Int, Any}
+    refcount::Int          # by how many Stmt objects referenced
+
+    function _Stmt(handle::StmtHandle)
+        stmt = new(handle, Dict{Int, Any}(), 0)
```

Comment:
why doesn't the refcount start at 1 here?

Reply:
Ok, I think I see how it's used later on; the refcount is how many `Stmt`s reference a `_Stmt`.

Hmmm, having reviewed the whole PR, the refcount stuff feels a little fragile/fiddly to me; can you restate the benefits of going with this approach of tracking `_Stmt`s apart from `Stmt` and the refcounting? I just want to make sure I understand what we gain/benefit for all this work, since it just seems like a lot of effort to just ensure we close things?

Reply:
> can you restate the benefits of going with this approach of tracking _Stmts apart from Stmt and the refcounting?

The idea is to confine the prepared statement handles within `DB` object and make 1-to-1 correspondence between internal `_Stmt` object and the prepared statement handle. E.g. even if the user will manage to create two `Stmt` objects that reference the same statement (e.g. by copying `Stmt`), there will be no double-close situation, because the handle will be closed only once (from `_Stmt` object). Also, when `DB` is closed, `Stmt` will be "aware" that the handle is no longer valid. I'm not an expert in SQLite3, so I don't know how bad it is to call its API methods with the bogus handle, but I assume it's better if we can intercept an error before the call.

And yes, it was also handy to have all prepared statement automatically tracked at one place, so that they could all be closed at once.
One can implement that on the user side, but it's much more efficient/less error-prone to do it in the package.

Reply:
As for refcounting itself, I'm not 100% sure about it either.
It's part of the solution to keep track of the `Stmt` scope when copying is possible.
If copying is forbidden (at least `copy` throws), it is not needed.
In that case double-close also should not happen, but there still needs
to be a mechanism of correctly closing statements upon DB close.